android-interop-testing/app/src/androidTest/java/io/grpc/android/integrationtest/TesterActivityTest.java

```diff
@@ -0,0 +1,58 @@
+package io.grpc.android.integrationtest;
+
+import android.os.Bundle;
+import android.test.ActivityInstrumentationTestCase2;
+import android.test.InstrumentationTestRunner;
+import junit.framework.Assert;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The Activity test of our App, used to trigger the integration test from commandline.
+ */
+public class TesterActivityTest extends ActivityInstrumentationTestCase2<TesterActivity> {
+
+  private String host;
+  private int port = 8030;
+  private String result;
+  private String serverHostOverride;
+  private String testCase;
+  private boolean useTls;
+  private boolean useTestCa;
+
+  public TesterActivityTest() {
+    super(TesterActivity.class);
+  }
+
+  public void setUp() throws Exception {
+    super.setUp();
+    getActivity();
+    Bundle args = ((InstrumentationTestRunner) getInstrumentation()).getArguments();
+    host = args.getString("server_host", "");
+    port = Integer.parseInt(args.getString("server_port", "0"));
+    assertTrue("Please specify server_host and server_port.", !host.isEmpty() && port != 0);
+    serverHostOverride = args.getString("server_host_override", null);
+    testCase = args.getString("test_case", "empty_unary");
+    useTls = Boolean.parseBoolean(args.getString("use_tls", "true"));
+    useTestCa = Boolean.parseBoolean(args.getString("use_test_ca", "false"));
+  }
+
+  public void testGrpc() throws Exception {
+    final CountDownLatch finished = new CountDownLatch(1);
+
+    new GrpcTestTask(testCase, host, port, serverHostOverride, useTls, useTestCa,
```

Comment:
I'm suggesting that we have a tester Activity that is given the Intent of which test to run. If the test is unsuccessful, then the Activity's result is something other than RESULT_OK. See http://developer.android.com/reference/android/app/Activity.html#StartingActivities . I would expect `am start -W` to return the same status code as the Activity's result.

For the issues:
 1. We will need to fix this. How do we expect our users to use the library if we don't know how?
   1. We don't actually need much of grpc-interop-testing if we are using nano proto on Android. Which part is being used?
 2. We do that same thing in our current tests. TestUtils.loadCert is used for this.
 3. Nothing listed is surprising, but yes, there are unknowns. (1) is _very_ important.

I have no issue postponing the Unit test for this PR.

For ProviderInstaller, I don't see any issue while reading the documentation. It does appear you are not following the documentation though. The Task should be calling installIfNeeded() (which blocks) or the task should only be created after onProviderInstalled(). The current code can race by starting the test before the provider is finished installing.